tests/scanner/audit/retention_rules_engine_test.py

```diff
@@ -0,0 +1,626 @@
+    res.append(data_creater.get_resource())
+
+    # bucket 4 with matches Storage Class
+    data_creater = frsd.FakeBucketDataCreater('bkt4', proj)
+    data_creater.AddLefecycleDict("Delete", 365, None, ["REGIONAL", "STANDARD", "DURABLE_REDUCED_AVAILABILITY", "NEARLINE", "COLDLINE"], None, None)
+    res.append(data_creater.get_resource())
+
+    # bucket 5 with numNewerVersions
+    data_creater = frsd.FakeBucketDataCreater('bkt5', proj)
+    data_creater.AddLefecycleDict("Delete", 365, None, None, 17, None)
+    res.append(data_creater.get_resource())
+
+    # bucket 6 with isLive
+    data_creater = frsd.FakeBucketDataCreater('bkt6', proj)
+    data_creater.AddLefecycleDict("Delete", 365, None, None, None, True)
+    res.append(data_creater.get_resource())
+
+    # bucket 7 with no "no match" violation
+    data_creater = frsd.FakeBucketDataCreater('bkt7', proj)
+    data_creater.AddLefecycleDict("Delete", 365, None, None, None, None)
+    data_creater.AddLefecycleDict("Delete", 201, None, None, None, True)
+    res.append(data_creater.get_resource())
+
+    # bucket 11 with nothing
+    data_creater = frsd.FakeBucketDataCreater('bkt11', proj)
+    data_creater.AddLefecycleDict("Delete", 365, None, None, None, None)
+    res.append(data_creater.get_resource())
+
+    # bucket 12 with created before
+    data_creater = frsd.FakeBucketDataCreater('bkt12', proj)
+    data_creater.AddLefecycleDict("Delete", 365, None, None, None, None)
+    data_creater.AddLefecycleDict("Delete", 365, "2018-01-01", None, None, None)
+    res.append(data_creater.get_resource())
+
+    # bucket 13 with matches Storage Class
+    data_creater = frsd.FakeBucketDataCreater('bkt13', proj)
+    data_creater.AddLefecycleDict("Delete", 365, None, None, None, None)
+    data_creater.AddLefecycleDict("Delete", 365, None, ["REGIONAL", "STANDARD", "DURABLE_REDUCED_AVAILABILITY", "NEARLINE", "COLDLINE"], None, None)
+    res.append(data_creater.get_resource())
+
+    # bucket 14 with numNewerVersions
+    data_creater = frsd.FakeBucketDataCreater('bkt14', proj)
+    data_creater.AddLefecycleDict("Delete", 365, None, None, None, None)
+    data_creater.AddLefecycleDict("Delete", 365, None, None, 17, None)
+    res.append(data_creater.get_resource())
+
+    # bucket 15 with is Live
+    data_creater = frsd.FakeBucketDataCreater('bkt15', proj)
+    data_creater.AddLefecycleDict("Delete", 365, None, None, None, None)
+    data_creater.AddLefecycleDict("Delete", 365, None, None, None, True)
+    res.append(data_creater.get_resource())
+
+    global _fake_bucket_list_0
+    _fake_bucket_list_0 = res
+    return res
+
+
+# TODO: More tests need to be added that cover the rule attributes and how they
+#    are evaluated
+class RetentionRulesEngineTest(ForsetiTestCase):
+    """Tests for the BigqueryRulesEngine."""
+
+    def setUp(self):
+        """Set up."""
+
+    @unittest.skipIf(do_not_test_old_cases, 'debug new test cases')
+    def test_retention_retrieve_1(self):
```

Comment:
It is useful when looking at test outputs to have more descriptive names, e.g.
test_invalid_rule_with_no_applies_to